### PR TITLE
Added r2 bucket lifecycle command (list, add, remove, set)

### PR DESCRIPTION
**Pull request by jonesphillip**
Added r2 bucket lifecycle command (list, add, remove, set)

---

- Tests
  - [ ] TODO (before merge)
  - [x] Tests included
  - [ ] Tests not necessary because:
- E2E Tests CI Job required? (Use "e2e" label or ask maintainer to run separately)
  - [ ] I don't know
  - [ ] Required
  - [x] Not required because: test changes covered by unit tests
- Public documentation
  - [ ] TODO (before merge)
  - [x] Cloudflare docs PR(s): https://github.com/cloudflare/cloudflare-docs/pull/18074
  - [ ] Documentation not necessary because:

### CODE REVIEW

**Review by irvinebroque — COMMENTED** [review]
awesome

File: packages/wrangler/src/r2/index.ts [hunk 1]
```diff
@@ -666,6 +667,37 @@ export function r2(r2Yargs: CommonYargsArgv, subHelp: SubHelp) {
...
 				}
 			);
+			r2BucketYargs.command(
+				"lifecycle",
+				"Manage lifecycle rules for an R2 bucket",
+				(lifecycleYargs) => {
+					return lifecycleYargs
+						.command(
+							"list <bucket>",
+							"List lifecycle rules for a bucket",
```
Comment (CarmenPopoviciu) [inline packages/wrangler/src/r2/index.ts:677]:
```suggestion
							"List lifecycle rules for an R2 bucket",
```
?

File: packages/wrangler/src/r2/lifecycle.ts [hunk 2]
```diff
@@ -0,0 +1,412 @@
...
+
+export function ListOptions(yargs: CommonYargsArgv) {
+	return yargs
+		.positional("bucket", {
+			describe: "The name of the R2 bucket to list lifecycle rules for",
+			type: "string",
+			demandOption: true,
+		})
+		.option("jurisdiction", {
+			describe: "The jurisdiction where the bucket exists",
```
Comment (CarmenPopoviciu) [inline packages/wrangler/src/r2/lifecycle.ts:31]: nit:
```suggestion
			describe: "The jurisdiction where the R2 bucket exists",
```

File: packages/wrangler/src/r2/lifecycle.ts [hunk 3]
```diff
@@ -0,0 +1,412 @@
...
+			requiresArg: true,
+		})
+		.option("abort-multipart-days", {
+			describe:
+				"Number of days after which incomplete multipart uploads are aborted",
+			type: "number",
+			requiresArg: true,
+		})
+		.option("jurisdiction", {
+			describe: "The jurisdiction where the bucket exists",
```
Comment (CarmenPopoviciu) [inline packages/wrangler/src/r2/lifecycle.ts:110]: nit
```suggestion
			describe: "The jurisdiction where the R2 bucket exists",
```

File: packages/wrangler/src/r2/lifecycle.ts [hunk 4]
```diff
@@ -0,0 +1,412 @@
...
+		enabled: true,
+		conditions: {},
+	};
+
+	if (!prefix && isInteractive) {
+		prefix = await prompt(
+			"Enter a prefix for the lifecycle rule (leave empty for all prefixes)"
+		);
+	}
+	if (prefix) {
```
Comment (CarmenPopoviciu) [inline packages/wrangler/src/r2/lifecycle.ts:163]: what should happen if there is still no prefix specified at this point?

File: packages/wrangler/src/dialogs.ts [hunk 5]
```diff
@@ -130,3 +130,28 @@ export async function select<Values extends string>(
...
 	return value;
 }
+
+export async function multiselect<Values extends string>(
+	text: string,
+	options: SelectOptions<Values>
+): Promise<Values[]> {
+	if (isNonInteractiveOrCI()) {
+		throw new NoDefaultValueProvided();
+	}
```
Comment (edmundhung) [inline packages/wrangler/src/dialogs.ts:140]: Nit: It would be nice if we can make it fallback to a defaultValue similar to the other prompts so it could be reused easily.

File: packages/wrangler/src/r2/helpers.ts [hunk 6]
```diff
@@ -792,6 +792,165 @@ export async function updateR2DevDomain(
...
+	const timestamp = date.getTime();
+	if (isNaN(timestamp)) {
+		return false;
+	}
+	const [year, month, day] = dateString.split("-").map(Number);
+	return (
+		date.getUTCFullYear() === year &&
+		date.getUTCMonth() + 1 === month &&
+		date.getUTCDate() === day
+	);
```
Comment (edmundhung) [inline packages/wrangler/src/r2/helpers.ts:943]: Why is this needed?

File: packages/wrangler/src/r2/lifecycle.ts [hunk 7]
```diff
@@ -0,0 +1,412 @@
...
+		throw new UserError("Must specify a rule ID.");
+	}
+
+	const newRule: LifecycleRule = {
+		id: id,
+		enabled: true,
+		conditions: {},
+	};
+
+	if (!prefix && isInteractive) {
```
Comment (edmundhung) [inline packages/wrangler/src/r2/lifecycle.ts:158]: Do you need to differentiate whether the user have specified a prefix or not?
```suggestion
	if (prefix === undefined && isInteractive) {
```

File: packages/wrangler/src/r2/lifecycle.ts [hunk 8]
```diff
@@ -0,0 +1,412 @@
...
+			demandOption: true,
+		})
+		.option("id", {
+			describe: "The unique identifier of the lifecycle rule to remove",
+			type: "string",
+			demandOption: true,
+			requiresArg: true,
+		})
+		.option("jurisdiction", {
+			describe: "The jurisdiction where the bucket exists",
```
Comment (CarmenPopoviciu) [inline packages/wrangler/src/r2/lifecycle.ts:302]: nit
```suggestion
			describe: "The jurisdiction where the R2 bucket exists",
```

File: packages/wrangler/src/r2/helpers.ts (hunk 6 above)
Comment (jonesphillip) [reply packages/wrangler/src/r2/helpers.ts:943]: When creating a Date object from input, it'll try to adjust the the date to something valid.
`new Date("02/31/2025")` becomes `3/3/2025` for example. This is just to check it matches user input.

File: packages/wrangler/src/r2/lifecycle.ts (hunk 4 above)
Comment (jonesphillip) [reply packages/wrangler/src/r2/lifecycle.ts]: If no actions are present in command line arguments and it's interactive, we'll prompt for prefix. If actions are present in command line arguments we will not prompt for prefix. However, in all cases we want will require confirmation if there is no prefix before adding a rule. This is important because if a prefix is omitted by accident, the action (potentially expiring objects, etc) will impact all objects in the bucket.

File: packages/wrangler/src/r2/lifecycle.ts (hunk 2 above)
Comment (jonesphillip) [reply packages/wrangler/src/r2/lifecycle.ts:32]: This is consistent with other jurisdiction option descriptions

File: packages/wrangler/src/r2/lifecycle.ts (hunk 3 above)
Comment (jonesphillip) [reply packages/wrangler/src/r2/lifecycle.ts:111]: This is consistent with other jurisdiction option descriptions

File: packages/wrangler/src/r2/lifecycle.ts (hunk 8 above)
Comment (jonesphillip) [reply packages/wrangler/src/r2/lifecycle.ts:322]: This is consistent with other jurisdiction option descriptions

File: packages/wrangler/src/r2/lifecycle.ts [hunk 9]
```diff
@@ -0,0 +1,412 @@
...
+			demandOption: true,
+		})
+		.option("file", {
+			describe: "Path to the JSON file containing lifecycle configuration",
+			type: "string",
+			demandOption: true,
+			requiresArg: true,
+		})
+		.option("jurisdiction", {
+			describe: "The jurisdiction where the bucket exists",
```
[the remark this answers is not in the document]
Comment (jonesphillip) [reply packages/wrangler/src/r2/lifecycle.ts:373]: This is consistent with other jurisdiction option descriptions

**Review by edmundhung — APPROVED** [review]
